sop-core/src/Data/SOP/NP.hs

```diff
@@ -130,12 +143,40 @@ import Data.SOP.Sing
 -- > K 0      :* K 1     :* Nil  ::  NP (K Int) '[ Char, Bool ]
 -- > Just 'x' :* Nothing :* Nil  ::  NP Maybe   '[ Char, Bool ]
 --
-data NP :: (k -> Type) -> [k] -> Type where
-  Nil  :: NP f '[]
-  (:*) :: f x -> NP f xs -> NP f (x ': xs)
+newtype NP (f :: k -> *) (xs :: [k]) = NP (V.Vector Any)
+
+-- | View on NP
+--
+-- This is only used internally, for the definition of the pattern synonyms.
+data ViewNP (f :: k -> *) (xs :: [k]) where
+  IsNil  :: ViewNP f '[]
+  IsCons :: f x -> NP f xs -> ViewNP f (x ': xs)
+
+-- | Construct 'ViewNP'
+--
+-- NOTE: 'V.unsafeTail' is O(1).
+viewNP :: NP f xs -> ViewNP f xs
+viewNP (NP xs)
+  | null xs   = unsafeCoerce $ IsNil
+  | otherwise = unsafeCoerce $ IsCons (unsafeCoerce (V.unsafeHead xs))
+                                      (NP (V.unsafeTail xs))
 
+pattern Nil :: forall f xs . () => (xs ~ '[]) => NP f xs
+pattern Nil <- (viewNP -> IsNil)
+  where
+    Nil = NP V.empty
+
+pattern (:*) :: forall f xs' . ()
+             => forall x xs . (xs' ~ (x ': xs)) => f x -> NP f xs -> NP f xs'
+pattern x :* xs <- (viewNP -> IsCons x xs)
+  where
+    x :* NP xs = NP (V.cons (unsafeCoerce x) xs)
```

Comment:
The should be a documentation warning here that the complexity of this for construction is O(n).